rest-api/dao/participant_counts_over_time_service.py

```diff
@@ -0,0 +1,156 @@
+
+class ParticipantCountsOverTimeService(ParticipantSummaryDao):
+
+  def __init__(self):
+    super(ParticipantSummaryDao, self).__init__(ParticipantSummary)
+
+
+  def get_filtered_results(self, start_date, end_date, filters, stratification='ENROLLMENT_STATUS'):
+    """Queries DB, returns results in format consumed by front-end
+
+    :param start_date: Start date string, e.g. '2018-01-01'
+    :param end_date: End date string, e.g. '2018-01-31'
+    :param filters: Objects representing filters specified in UI
+    :param stratification: How to stratify (layer) results, as in a stacked bar chart
+    :return: Filtered, stratified results by date
+    """
+
+    start_date = start_date.replace('-', '')
+    end_date = end_date.replace('-', '')
+
+    filters_sql = self.get_facets_sql(filters)
+
+    if stratification == 'TOTAL':
+      strata = ['TOTAL']
+      sql = """
+        SELECT
+            SUM(ps_sum.cnt * (ps_sum.day <= calendar.day)) registered_count,
+            calendar.day start_date
+        FROM calendar,
+        (SELECT COUNT(*) cnt, DATE(ps.sign_up_time) day
+        FROM participant_summary ps
+        GROUP BY day) ps_sum
+        WHERE calendar.day >= :start_date
+        AND calendar.day <= :end_date
+        GROUP BY calendar.day
+        ORDER BY calendar.day;
+      """
+    elif stratification == 'ENROLLMENT_STATUS':
+      strata = [str(EnrollmentStatus(val)) for val in EnrollmentStatus]
+      sql = """
+      SELECT
+         SUM(registered_cnt * (cnt_day <= calendar.day)) registered_participants,
+         SUM(member_cnt * (cnt_day <= calendar.day)) member_participants,
+         SUM(full_cnt * (cnt_day <= calendar.day)) full_participants,
+         calendar.day
+       FROM
+       (SELECT c2.day cnt_day,
+               registered.cnt registered_cnt,
+               member.cnt member_cnt,
+               full.cnt full_cnt
+            FROM calendar c2
+            LEFT OUTER JOIN
+            (SELECT COUNT(*) cnt, DATE(ps.consent_for_study_enrollment_time) day
+               FROM participant_summary ps
+              WHERE %(filters)s
+              GROUP BY day) registered
+             ON c2.day = registered.day
+           LEFT OUTER JOIN
+            (SELECT COUNT(*) cnt,
+                    DATE(ps.consent_for_electronic_health_records_time) day
+               FROM participant_summary ps
+              WHERE %(filters)s
+           GROUP BY day) member
+             ON c2.day = member.day
+           LEFT OUTER JOIN
+            (SELECT COUNT(*) cnt,
+             DATE(CASE WHEN enrollment_status = 3 THEN
+                   GREATEST(consent_for_electronic_health_records,
+                            questionnaire_on_the_basics_time,
+                            questionnaire_on_lifestyle_time,
+                            questionnaire_on_overall_health_time,
+                            physical_measurements_time,
+                            CASE WHEN sample_status_1ed04_time IS NOT NULL
```

Comment:
This works OK but hardcodes the baseline questionnaires and the fact that 1ED04 and 1SAL are the samples needed to be a full participant into the SQL. Previously this was done with config:

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/config/base_config.json#L29

For the samples, MySQL doesn't make it easy to do the LEAST of N nullable values, so this is probably OK... but at a minimum it means we should add some comments about it in here.

For the baseline questionnaires, we might want to use the config to generate the GREATEST statement instead. Or make sure to document this if we're not going to.

Also, GREATEST and LEAST don't work in SQLite. So to unit test this you're going to need to add macros for them, e.g.:

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/dao/database_utils.py#L50

Reply:
I've documented those comments in the code.

For the GREATEST and LEAST, my plan to implement those macros in my pull request for unit tests.  (In our early meetings on this, you recommended sending four PRs, latter two for tests of API and service, so my plan is to do the macros in my next PR.)